KAFKA-18613: Unit tests for usage of incorrect RPCs

lucasbru opened the pull request:
In the GroupMetadataManager, we may call an RPC on an incorrect group type. This adds unit tests to validate the behavior when an RPC is used on an incorrect group type.

Reviewers: Bill Bejeck <bill@confluent.io>

lucasbru committed:
KAFKA-18613: Unit tests for usage of incorrect RPCs / group type pairs

In the GroupMetadataManager, we may call an RPC on an incorrect group
type. This adds unit tests to validate the behavior when an RPC is
used on an incorrect group type.

lucasbru commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -741,6 +741,7 @@ public List<StreamsGroupDescribeResponseData.DescribedGroup> streamsGroupDescrib
                 describedGroups.add(new StreamsGroupDescribeResponseData.DescribedGroup()
                     .setGroupId(groupId)
                     .setErrorCode(Errors.GROUP_ID_NOT_FOUND.code())
+                    .setErrorMessage(exception.getMessage())
```
This was actually inconsistent to other group types

lucasbru commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/taskassignor/StickyTaskAssignorTest.java:
```diff
@@ -18,6 +18,7 @@
 package org.apache.kafka.coordinator.group.taskassignor;
 
 import org.apache.kafka.coordinator.group.GroupCoordinatorConfig;
+
```
spotlessApply

lucasbru commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -9373,7 +9375,8 @@ public void testStreamsGroupDescribeBeforeAndAfterCommittingOffset() {
             .setGroupId(streamsGroupId)
-            .setErrorCode(Errors.GROUP_ID_NOT_FOUND.code());
+            .setErrorCode(Errors.GROUP_ID_NOT_FOUND.code())
+            .setErrorMessage("Group " + streamsGroupId + " not found.");
```
Updates behavior according to the above change to production code.

lucasbru commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -15673,6 +15676,349 @@ public void testReplayStreamsGroupTopologyTombstone() {
     }
 
+    @Test
+    public void testConsumerGroupHeartbeatOnStreamsGroup() {
```
The test are pretty repetitive - like the ones for the other group types. I considered saving some LOC (by using private helper methods, and parametrized tests), but kept them at way, since it's going to be easier to refactor if each test is self-contained.

lucasbru commented:
@bbejeck is this still on your list?

bbejeck reviewed (approved):
Thanks for the PR @lucasbru - overall LGTM.  There does not seem to be too much variance in creating a majority of the `GroupMetadataManagerTestContext` object, maybe one helper method could help in this case.  But that's up to you.

lucasbru committed:
KAFKA-18613: Improve test coverage for missing topics

Tests for joining with missing source topics,
internal topics, incorrectly partitioned topics,
and stale topologies.

lucasbru commented:
I retargeted this to trunk (I thought I had already). The tests weren't changed.

lucasbru committed:
comments

lucasbru committed:
Merge remote-tracking branch 'apache/trunk' into kip1071pr/gmm_tests
